### PR TITLE
Add search filter to app list

### LINKED ISSUE #1617: Add search parameter to AppRegistryController.list()
Currently the `AppRegistryController` does not have the ability to filter the results. We should add the `search` parameter similar to the `StreamDefinitionController`.

### PR DESCRIPTION
**Pull request by jvalkeal**
- Add simple search match by apps name. Can't
  really filter by address as it would give
  unrealistic results together with name.
- Fix some issues with pagination i.e. when page
  which doesn't exist is requested. Also make paging
  work with filtered set.
- Fixes #1617

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/AppRegistryControllerTests.java [hunk 1]
```diff
@@ -154,6 +154,24 @@ public void testListApplicationsByType() throws Exception {
 	}
 
 	@Test
+	public void testListApplicationsBySearch() throws Exception {
+		mockMvc.perform(get("/apps?search=timestamp").accept(MediaType.APPLICATION_JSON)).andDo(print()).andExpect(status().isOk())
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/AppRegistryControllerTests.java]: we can remove the print().  I can do that on merge.

Comment (jvalkeal) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/AppRegistryControllerTests.java]: Indeed there is a lot of other tests to log these as well. Feel free to remove those as well as we don't need to be so noisy.

**Comment by cppwfs** [comment]
LGTM

**Comment by cppwfs** [comment]
Polished, rebased, squashed, merged.

**Pull request closed without merging**